crates/core_simd/src/ops/splat.rs

```diff
@@ -0,0 +1,74 @@
+//! This module implements "auto-splats" for operators.
+//! These should be handled with care due to the risks to type inference.
+//! We require for every autosplat that:
+//! - Simd<T, _> implements the operation for itself, returning Simd<T, _>
+//! - T implements the same operation for T, returning T
+//!
+//! The second bound is not strictly necessary, but enforces symmetry
+//! between a vector operation and repeating the scalar equivalent.
+use super::*;
+
+macro_rules! autosplat_rhs {
+    ($(impl<T, const LANES: usize> $op:ident<$operand:ty> for Simd<$scalar:ty, LANES> {
+        fn $call:ident
+    })*) => {
+        $(impl<T, const LANES: usize> $op<$operand> for Simd<$scalar, LANES>
+        where
+            Self: $op<Self, Output=Self>,
+            $operand: SimdElement + $op<T, Output=T>,
```

Comment:
The op bound on the scalar is unnecessary

Reply:
> //! The second bound is not strictly necessary, but enforces symmetry
//! between a vector operation and repeating the scalar equivalent.

I would like to retain this one at the moment to keep us from making the kinds of mistakes that I fixed in 1ce1c645cf27c4acdefe6ec8a11d1f0491954a99, 949f71c0dc8716e285c164eae225db2d18333f0d, and c2f59483f96cf1ab1e92cf10e0f9094432a8374c... essentially, to prevent anything from sneaking past us. I don't expect it to remain in the final implementation and it's completely forward-compatible to **relax** bounds.

Reply:
There is no intrinsic used here, only the trait required by the bound on Self is used.

Reply:
The previous version did use the intrinsic directly, however, and in the future it might do so again.

Reply:
It's not, however, about the intrinsic. It's about whether the operator works normally on the scalar at all. For some of the scalar types you've suggested we might want to include, **it does not**, even though they are also numerical types with an in-range repr and thus **would** be valid to monomorphize. But some do implement **some** operators.

If we are including them in our discussions, then I insist on overconstraining our implementation to discourage accidentally implementing operators for them where those have not been implemented, until we have actually implemented them.

Reply:
Well, I still disagree that Add on the scalar means the add intrinsic works on the vector, then.  All it means is that you can call `Add::add` on it, and nothing more.

But in this case, seeing as it's the splatting case, by definition it's bound by whether you can add two vectors of the same type.

Reply:
Precisely, the bound is separately restated for each case.

Reply:
This is all especially desirable so if you want the macro implementation on the base types to not have bounds, where they would admittedly be even more superfluous. But I want to rely a **lot** less on human memory and human eyes in code review. The alternative is writing a machine to verify we don't implement an operator for a vector if its scalar doesn't, and we have a compiler right in front of us to do that.

Reply:
Since relaxing bounds is perfectly forward compatible, we could stabilize this version tomorrow and it would **still** be possible to remove bounds... but not add them.

Reply:
Updating this to fix rust-lang/rust#90904, which eliminates this discussion.